[SILOptimizer] Added MandatoryCombiner to passes.

nate-chandler opened the pull request:
The mandatory combiner is added both to the Onone pass and also to the performance pass.

nate-chandler commented:
@swift-ci please smoke test

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp:
```diff
@@ -143,20 +143,25 @@ struct MandatoryCombiner : SILFunctionTransform {
-          // deleted.  Instead, it will advance it to the next *remaining*
-          // instruction after the previous instruction as desired.
-          ii = next_or_default(ii, /*end*/block.end(), /*defaultIter*/block.begin());
+          // The current instruction has been deleted.  The iterator has already
```
Can you move any changes that are not directly related to adding this to the pass pipeline into a different PR (or at least a different commit)?

nate-chandler replied:
Yup!  Will split into separate commit.

nate-chandler replied:
https://github.com/apple/swift/pull/26924/commits/e4cee7fdfaac4051b8b75f7f042f93d20f22bb1e

gottesmm commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -658,6 +660,8 @@ SILPassPipelinePlan::getOnonePassPipeline(const SILOptions &Options) {
   P.startPipeline("Serialization");
   P.addSerializeSILPass();
 
+  P.addMandatoryCombiner();
```
I think we want to add this before serialization.

nate-chandler replied:
Fixed.

gottesmm commented on lib/SILOptimizer/PassManager/PassPipeline.cpp:
```diff
@@ -393,6 +393,8 @@ static void addPerfEarlyModulePassPipeline(SILPassPipelinePlan &P) {
   // we do not spend time optimizing them.
   P.addDeadFunctionElimination();
 
+  P.addMandatoryCombiner();
```
I am fine with this here, but this may make more sense in the SILOptPreparePassPipeline pipeline. Your thoughts?

nate-chandler replied:
It happening earlier seems only for the good.  Fixed.

gottesmm reviewed (approved):
LGTM

nate-chandler commented:
@swift-ci please test and merge

gottesmm commented:
@swift-ci please test and merge

nate-chandler committed:
[SILOptimizer] Added MandatoryCombiner to passes.

The mandatory combiner is added both to the Onone pass and also to the
performance pass.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please ASAN test

nate-chandler commented:
@swift-ci please test and merge

nate-chandler commented:
@swift-ci please test and merge
